HAST-184: Add Zynq documentation (#4)
Lombiq/Hastlayer-Hardware-Framework---Vitis: Docs/ZynqVivadoXsa.md

# Building Vivado XSA (Xilinx Shell Archive) file for Hastlayer (Trenz TE0715-04-30-1C module)

To start with the Hastlayer on the Trenz TE0715-04-30-1C module you first need to build a Vivado XSA file containing the appropriate Zynq Processing System (ARM CPU) and FPGA fabric configuration. This document describes the required steps.

## Install Vitis Core Development Kit 2020.2

> 

Reviewer: ```suggestion

> ℹ️ If you are using the Centos 7 Docker image you already have the development kit and can skip this section.

```

Download the Vitis Core Development Kit 2020.2 installer (Vivado included) from Xilinx website: (https://www.xilinx.com/support/download/index.html/content/xilinx/en/downloadNav/vitis/2020-2.html) and install it. 

![Download Vitis Core Development Kit Installer](Images/VitisDownloadInstaller.png)

You can install Vitis and Vivado as root user in the default /tools/Xilinx folder.

Before using the Vivado tools you first have to set up the environment:

```
source /tools/Xilinx/Vitis/2020.2/settings64.sh
```

## Downloading and configuring the reference design for TE0715-04-30-1C module

Create an empty directory, then download and extract the reference design from Trenz Electronic:

```
mkdir ${HOME}/trenz_te0715_04_30_1c
cd ${HOME}/trenz_te0715_04_30_1c
mkdir vivado
cd vivado

> wget https://shop.trenz-electronic.de/trenzdownloads/Trenz_Electronic/Modules_and_Module_Carriers/4x5/TE0715/Reference_Design/2020.2/test_board/TE0715-test_board-vivado_2020.2-build_5_20210611100936.zip

Reviewer: Ez a link halott. Úgy látom amikor frissítik (legutoljára december 16-án) akkor a korábbit letörlik ezért nem lesz jó így hotlinkelni.
Helyette le kéne írni, hogy innen kell letölteni:
https://shop.trenz-electronic.de/Download/?path=Trenz_Electronic/Modules_and_Module_Carriers/4x5/TE0715/Reference_Design/2020.2

Reviewer: Erre is válaszolj majd, és kérlek ne nyomj a <kbd>Resolve conversation</kbd>-re. Az a reviewer feladata. (Persze ha nyomsz <kbd>Commit suggestion</kbd>-t akkor a rendszer magátol resolve-olja, az rendben van.)

> unzip TE0715-test_board-vivado_2020.2-build_5_20210611100936.zip

Reviewer: ```suggestion
unzip TE0715-test_board-vivado_2020.2-*.zip
```

```

## Configure the reference design (command line part)

```
cd ${HOME}/trenz_te0715_04_30_1c/vivado/test_board
source _create_linux_setup.sh
```

> In the command line menu first press "0" (Module selection guide, project creation...). Wait until the module list apears. Then type in "AN", ENTER, "TE0715-04-30-1C", ENTER to select the apropriate module. Then answer "Y",  and press "2" at the end to create the Vivado project.

Reviewer: ```suggestion
In the command line menu: 
- Press "0". (Module selection guide, project creation...)
- Wait until the module list apears.
- Type in `AN`, then press <kbd>enter</kbd>.
- Type `TE0715-04-30-1C`, then press <kbd>enter</kbd> to select the apropriate module.
- Answer `Y`, and then `2` at the end to create the Vivado project.
```

> In the command line menu first press "0" (Module selection guide, project creation...). Wait until the module list apears. Then type in "AN", ENTER, "TE0715-04-30-1C", ENTER to select the apropriate module. Then answer "Y",  and press "2" at the end to create the Vivado project.

Reviewer: Ehhez miért kell GUI, és hogy tudom letiltani, hogy ne kelljen?
```
Start GUI...all other messages will be print inside the GUI TCL console of Vivado
Warning: This program is an suid-root program or is being run by the root user.
The full text of the error or warning message cannot be safely formatted
in this environment. You may get a more descriptive message by running the
program as a non-root user or by removing the suid bit on the executable.
xterm: Xt error: Can't open display: %s
ERROR:(TE) Script (TE::INIT::run_board_selection) failed: ERROR: [Common 17-1374] Display test failed. Please check DISPLAY environment variable and X server permissions for remote display.
.
ERROR:(TE) Script (TE::main) failed: .
INFO: [Common 17-206] Exiting Vivado at Fri Dec 24 14:23:52 2021...
```

Wait until the Vivado GUI is done with the initial block design creation.

## Configure the reference design for Hastlayer (Vivado GUI part)

Change the project type to Vitis extensible platform on Project Settings - General tab:

![Enable Vitis Platform](Images/ZynqXsaVitisPlatform.png)

Import the external constraint files into the project to be able to export the XSA file. Right-click on Constraint(5) filegroup and choose "Add Sources..."

![Import constraint step 1](Images/ZynqXsaImportConstraints1.png)

Be sure that "Copy constraints files into project" settings are checked, and press "Finish".

![Import constraint step 2](Images/ZynqXsaImportConstraints2.png)

### Configure the "ZYNQ7 Processing System"

Set FCLK_CLK0 to 150 MHz:

![Set FCLK_CLK0](Images/ZynqXsaPsFclk0.png)

> Enable iinterrupt requests from fabric:

Reviewer: ```suggestion
* Go to Interrupts on the side menu.
* Tick the Fabric Interrupts.
* Tick Fabric Interrupts > PL_PS Interrupt Ports > IRQ_F2P.
* Click Ok.
```

![Enable IRQ_F2P](Images/ZynqXsaPsIrqF2P.png)

### Extend the block design with the Concat and Reset IP

> Right-click on the block design and add "Concat" and "Processor System Reset" IPs.

Reviewer: ```suggestion
* Right-click on the block design.
* Select Add IP.
* Type "Concat" into the search bar and click on the result.
* Do the same for "Processor System Reset" too.
```

Make connection between:

> xlconcat_1:dout[1:0] -> PS:IRQ_F2P[1:0]
> 
> PS:FCLK_CLK0 -> proc_sys_reset_0:slowest_sync_clk
> PS:FCLK_RESET0_N -> proc_sys_reset_0:ext_reset_in

Reviewer: ```suggestion
* `xlconcat_1:dout[1:0]` -> `processing_system7_0:IRQ_F2P[1:0]`
* `processing_system7_0:FCLK_CLK0` -> `proc_sys_reset_0:slowest_sync_clk`
* `processing_system7_0:FCLK_RESET0_N` -> `proc_sys_reset_0:ext_reset_in`
```

![Concat and Reset](Images/ZynqXsaConcatReset.png)

> ###Configure the "Platform Setup" pages

Reviewer: ```suggestion
### Configure the "Platform Setup" pages
```

Enable M_AXI_GP0 and S_AXI_HP0:

![VitisXsaPlatformAxi](Images/ZynqXsaPlatformAxi.png)

Enable FCLK_CLK0:

![VitisXsaPlatformClock](Images/ZynqXsaPlatformClock.png)

Enable interrupt In0 .. In7:

![VitisXsaPlatformInterrupt](Images/ZynqXsaPlatformInterrupt.png)

> 

Reviewer: ```suggestion

Save and close the Block Design tab.

```

## Export the XSA file

> Build the bitfile before exporting XSA by pressing the "Generate Bitstream" in Flow Navigator.
> 
> After successfully building the bitfile press "Export Platform" in Flow Navigator. The platform type should be hardware only / pre-synthesis / include bitfile.

Reviewer: ```suggestion
* Click PROGRAM AND DEBUG > Generate Bitstream on the sidebar.
* Press Yes on the prompt and Ok on the dialog box.
* You need to wait for synthesis and implementation. You can track the progress on the Project Summary panel.
* Eventually a dialog will pop up telling you that the implementation has completed. You can close it.
* Click IP INTEGRATOR > Export Platform on the sidebar.
* Click Next until you reach the Platform State where you need to tick the Include bitstream box. Then click Next until the end and then Finish.
   * Note that you can see the path on the Output File page, it should be the same as below.
```

> After successfully building the bitfile press "Export Platform" in Flow Navigator. The platform type should be hardware only / pre-synthesis / include bitfile.

Reviewer: Ezt dobja, amikor kész? Kimaradt egy lépés vagy félre értettem valamit?
![image](https://user-images.githubusercontent.com/4007293/148559053-b0e79ea8-fd15-4a37-a123-fcaa65659a39.png)

# Building Vivado XSA (Xilinx Shell Archive) file for Hastlayer (Trenz TE0715-04-30-1C module)

To start with the Hastlayer on the Trenz TE0715-04-30-1C module you first need to build a Vivado XSA file containing the appropriate Zynq Processing System (ARM CPU) and FPGA fabric configuration. This document describes the required steps.

## Install Vitis Core Development Kit 2020.2

> ℹ️ If you are using the Centos 7 Docker image you already have the development kit and can skip this section.

Download the Vitis Core Development Kit 2020.2 installer (Vivado included) from Xilinx website: (https://www.xilinx.com/support/download/index.html/content/xilinx/en/downloadNav/vitis/2020-2.html) and install it. 

![Download Vitis Core Development Kit Installer](Images/VitisDownloadInstaller.png)

You can install Vitis and Vivado as root user in the default /tools/Xilinx folder.

Before using the Vivado tools you first have to set up the environment:

```
source /tools/Xilinx/Vitis/2020.2/settings64.sh
```

## Downloading and configuring the reference design for TE0715-04-30-1C module

> Create an empty directory, then download and extract the reference design from Trenz Electronic:
> 
> ```
> mkdir ${HOME}/trenz_te0715_04_30_1c
> cd ${HOME}/trenz_te0715_04_30_1c
> mkdir vivado
> cd vivado
> wget https://shop.trenz-electronic.de/trenzdownloads/Trenz_Electronic/Modules_and_Module_Carriers/4x5/TE0715/Reference_Design/2020.2/test_board/TE0715-test_board-vivado_2020.2-build_5_20210611100936.zip
> unzip TE0715-test_board-vivado_2020.2-*.zip
> ```

Reviewer 2: @Piedone Újraírtam ezt a részt, ha szerinted ok akkor mehet.

```suggestion
Create an empty directory, then download and extract the reference design from Trenz Electronic into it. As of 2024 the latest file can be found [here](https://shop.trenz-electronic.de/trenzdownloads/Trenz_Electronic/Modules_and_Module_Carriers/4x5/TE0715/Reference_Design/2022.2/test_board/TE0715-test_board-vivado_2022.2-build_2_20230705115102.zip). But if that link is outdated, please follow these instructions:

1. Go to the [TE0715 Resources page](https://wiki.trenz-electronic.de/display/PD/TE0715+Resources) in the Trenz Electronic Wiki.
2. Click the _Download area_ link.
3. Follow this path: Reference_Design > 2022.2 > test_board.
4. Download the archive in the _Reference Design - Source Code and Configuration Files_ section.
```

Reviewer 3: `>` helyett `→`-t használnék, de amúgy jó lesz.

Reviewer 2: Na, akkor csak érdemes volt rákérdezni! :D
